modyn/trainer_server/internal/trainer/pytorch_trainer.py

```diff
@@ -347,6 +373,82 @@ def train(self) -> None:  # pylint: disable=too-many-locals, too-many-branches
 
         self._info("Training complete!")
 
+    def prepare_data(self, batch: tuple) -> tuple[list, torch.Tensor, Union[torch.Tensor, dict]]:
+        sample_ids = batch[0]
+        if isinstance(sample_ids, torch.Tensor):
+            sample_ids = sample_ids.tolist()
+        elif isinstance(sample_ids, tuple):
+            sample_ids = list(sample_ids)
+
+        assert isinstance(sample_ids, list), "Cannot parse result from DataLoader"
+
+        if self._label_tranformer_function is None:
+            target = batch[2].to(self._device)
+        else:
+            target = self._label_tranformer_function(batch[2]).to(self._device)
+
+        data: Union[torch.Tensor, dict]
+        if isinstance(batch[1], torch.Tensor):
+            data = batch[1].to(self._device)
+        elif isinstance(batch[1], dict):
+            data: dict[str, torch.Tensor] = {}  # type: ignore[no-redef]
+            for name, tensor in batch[1].items():
+                data[name] = tensor.to(self._device)
+        else:
+            raise ValueError(
+                "The format of the data provided is not supported in modyn. "
+                "Please use either torch tensors or dict[str, torch.Tensor]"
+            )
+
+        return sample_ids, target, data
+
+    def sample_data(self) -> None:
+        """
+        Function to score every datapoint in the current dataset and sample a fraction of it
+        Used for downsampling strategies in sample_then_batch mode
+
+        """
+        assert self._downsampler is not None
+        assert self._downsampling_mode == DownsamplingMode.SAMPLE_THEN_BATCH
+
+        # keys must be taken from the selector.
+        # This operation is needed only when we sample several times (otherwise the source is already the selector)
+        self._train_dataloader.dataset.switch_to_selector_key_source()
+
+        number_of_batches = 0
+
+        self._downsampler.init_downsampler()
+
+        sample_ids_order = []
```

Comment:
Not sure from the name what it is used for. Could you maybe add a comment here?

Reply:
So, it's the same problem I was trying to explain a few days ago so I'll spend a little more time on it now.
Let's imagine we have the following sample ids [101, 102,103, 115, 123, 107] and a batch size of 3.

At the first iteration, the downsampler receives the output of the model when there is as input [101, 102, 103] and saves the score (gradnorm/loss/whatever) in a list/matrix. So 101 is saved at index 0, 102 at index 1 and so on. Same for the second batch, so the last element (107) has index 5. For example, if we use loss we can have something like [1.21, 0.76, 0.54, 5.23, 6.12, 1.11] and then sample from this list (for CRAIG we have a matrix).

The downsampler selects data based on its data structure (so indexes from 0 to 5) and for example returns [3, 4, 1] (right now the downsampler is not aware of the sample ids)

Consequently we need to maintain the reverse mapping and know that [3] corresponds to key 115, that [4] is key 123 and [1] is key 102. We can do this operation either inside the downsampler or outside (as it is done now), I don't know which one is better.

Reply:
I think as this is logic for the downsampler itself, it would be best if this can be "hidden" inside the downsampler. Furthermore, could we rename the variable to index_sampleid_map? I think map would be the more canonical name. And probably, you can just copy some sentences from here as a comment, so we preserve the knowledge for future generations :D Sounds good in general.